PXC-4076 8.0.31-Documentation for GCache and Write-Set Cache encryption (8.0) (#44)
percona/pxc-docs: docs/management/gcache-write-set-cache-encryption.md

> # GCache and Write-Set cache encryption

Reviewer: Just to be clear: 
1. GCache
2. Write-Set = Record-Set
These are two separate things

Author: Updated.

The feature is a [tech preview](../glossary.md#tech-preview). Before using this feature in production, we recommend that you test restoring production from physical backups in your environment, and also use the alternative backup method for redundancy.

> ## GCache encryption Master Key

Reviewer: Encryption Master Key

Author: Rewritten

> GCache encrypts with a File Key. Creating the galera.gcache file or a full reset generates the File Key. The File Key is encrypted with the Master Key. 

Reviewer: GCache and Write-Set cache files created by Galera are encrypted with File Key if this feature is enabled. Only Ring Buffer file of GCache is non-volatile (it survives the server restart), so only for this file, the File Key has to be retained. It is stored after encrypting with Master Key which is stored in the keyring.
For GCache off-pages and Write-Set cache files, File Key is not stored anywhere.

Author: Rewritten

!!! admonition "See also"

    For more information, see [Understanding GCache and Record-set Cache](../manual/gcache_record-set_cache_difference.md), and the Percona Database Performance Blog: [All you need to know about GCache](https://www.percona.com/blog/2016/11/16/all-you-need-to-know-about-gcache-galera-cache/)

> GCache has a RingBuffer on-disk file to manage write-sets. The file is pre-created when the server starts. This file contains a 1KB preamble to hold key-value pairs. The Master Key id and the encrypted File Key are stored in the GCache Ring preamble. The preamble is not encrypted because this section does not store sensitive data. 

Reviewer: I'm not sure if we need this information about the preamble, its format, etc. This is the internal thing to Galera, user has nothing to set there (only curious user can look inside and see), but this could be a binary format of the preamble as well and it would have no inpact to the user.

Author: Removed.

??? example "Sample preamble key-value pairs"

    ```text

    Version: 2
    GID: 3afaa71d-6665-11ed-98de-2aba4aabc65e
    synced: 0
    enc_version: 1
    enc_encrypted: 1
    enc_mk_id: 3
    enc_mk_const_id: 3ad045a2-6665-11ed-a49d-cb7b9d88753f
    enc_mk_uuid: 3ad04c8e-6665-11ed-a947-c7e346da147f
    enc_fk_id: S4hRiibUje4v5GSQ7a+uuS6NBBX9+230nsPHeAXH43k=
    enc_crc: 279433530

    ```

### Key descriptions

The following table describes the encryption keys defined in the preamble. All other keys in the preamble are not related to encryption.

| Key | Description |
|---|---|
| `enc_version` | The encryption version |
| `enc_encrypted` | If the GCache is encrypted or not |
| `enc_mk_id` | A part of the Master Key ID. Rotating the Master Key increments the sequence number. |

> | `enc_mk_const_id` | A part of the Master Key ID, a constant Universally unique identifier (UUID). This option remains constant for the duration of the galera.gcache file. This option simplifies matching the keys inside the keyring to the instance that generated the keys. Deleting the galera.gcache changes this option. |

Reviewer: This is not the option. User cannot change it. It is a key/identifier. I would rather say:
It remains constant...
It simplifies matching the Master Key inside the keyring...
...changes the value of this key.

Author: Rewritten

> | `enc_mk_uuid` | The first Master Key or if Galera detects that the preamble is inconsistent, which causes a full GCache reset and a new Master Key is required, generates this UUID. This change avoids reuse of existing keys, which may be compromised, from the keyring.  |

Reviewer: `This change avoids reuse of existing keys, which may be compromised, from the keyring.`
Implementation detail. Remove

Author: Removed.

> | `enc_fk_id` | The File Key ID used for GCache ring buffer encryption. Encrypted with the Master Key and stored in base64 encoding. |

Reviewer: Just: File Key encrypted with Master Key

Author: Rewritten.

| `enc_crc` | The cyclic redundancy check (CRC) calculated from all encryption-related keys. |

### Controlling encryption

Encryption is controlled using the wsrep_provider_options. 

| Variable name | Default value | Allowed values |
|---|---|---|
| [`gcache.encryption`](#gcacheencryption) | off | on/off |
| [`gcache.encryption_cache_page_size`](#gcacheencryption_cache_page_size) | 32KB | 2-512 |
| [`gcache.encryption_cache_size`](#gcacheencryption_cache_size) | 16MB | 2 - 512  |
| [`allocator.disk_pages_encryption`](#allocatordisk_pages_encryption) | off | on/off |
| [`allocator.encryption_cache_page_size`](#allocatorencryption_cache_page_size) | 32KB |  |
| [`allocator.encryption_cache_size`](#allocatorencryption_cache_size) | 16MB |  |

## Variable descriptions

> The following sections describe the variables related to Record-Set cache encryption.

Reviewer: Let's split it into 2 parts:
1. ...related to GCache encryption
2. ...related to Write-Set cache encryption

Author: Added.

All variables are read-only.

### gcache.encryption

Enable or disable GCache encryption.

### gcache.encryption_cache_page_size

> The size of the encryption cache for Record-Set pages. The value must be multiple of the CPU page size (typically 4kB). If the value is not, the server reports an error and stops.

Reviewer: The size of the GCache encryption cache page.

Author: Updated.

> The maximum limit is 512. The minimum limit is 2.

Reviewer: Remove

Author: Done.

### gcache.encryption_cache_size

> The size of the encryption cache for GCache on-disk files. If `gcache.encryption=yes`, one encryption cache is associated with the GCache Ringbuffer file. Create an on-disk page and an encryption cache for that page is also created.

Reviewer: Encryption uses so-called 'encryption cache'. Each encrypted file has its own encryption cache associated. Encryption cache consists of pages.
Page size is configured via `gcache.encryption_cache_page_size`
Encryption cache size is configured by `gcache.encryption_cache_size`

The maximum limit of pages in encryption cache is 512. The minimum limit is 2.

Author: Rewritten.

> The value cannot be smaller than the CPU page size, typically 4KB. If the value is smaller, the server reports an error and stops. We recommend a large cache since an extremely small encryption cache causes frequent encrypted pages swapping and degrade performance. 

Reviewer: The value cannot be smaller than 2 encryption cache pages. If the value is smaller it will be rounded up.
As the limit of pages allocated is 512, this value is a hint. If too big, the size will be rounded down to 512 x gcache.encryption_cache_page_size.

Author: Rewritten.

The `encryption_cache_size` should be a multiple of `encryption_cache_page_size`. If the value is not, the actual cache size is rounded up.

The `encryption_cache_size` maximum is 512 allocation pages in the encryption cache. If the cache size must increase, you must also increase the `allocator.encryption_cache_page_size`.

### allocator.disk_pages_encryption

Enable or disable the allocator's disk pages encryption.

> ### allocator.encryption_cache_page_size

Reviewer: Here we start 2nd section: related to Write-Set cache encryption

Parameters are similar to ones in GCache section above, description is similar, the only difference is that now we are talking about Write-Set cahce not GCache

Author: Added.

# GCache encryption and Write-Set cache encryption

These features are [tech preview](../glossary.md#tech-preview). Before using these features in production, we recommend that you test restoring production from physical backups in your environment, and also use the alternative backup method for redundancy.

> ## GCache and Write-Set

Reviewer: Some left-over?

Author: Removed.

In Percona XtraDB Cluster, 

## GCache and Write-Set cache encryption

Enabling this feature encrypts the Galera GCache and Write-Set cache files with a File Key.  
 

>   GCache has a RingBuffer on-disk file to manage write-sets. The keyring only stores the Master Key and the File Key used by the Ring Buffer file. The Ring Buffer file of GCache is non-volatile, which means this file survives a restart. The File Key is not stored for GCache off-pages and Write-Set cache files.

Reviewer: The keyring stores only MK.
The keyring only stores the Master Key which is used to encrypt the File Key used by the Ring Buffer file. The encrypted File Key is stored in the Ring Buffer's preamble.

Author: Rewritten.

### Controlling encryption

Encryption is controlled using the wsrep_provider_options. 

| Variable name | Default value | Allowed values |
|---|---|---|
| [`gcache.encryption`](#gcacheencryption) | off | on/off |
| [`gcache.encryption_cache_page_size`](#gcacheencryption_cache_page_size) | 32KB | 2-512 |
| [`gcache.encryption_cache_size`](#gcacheencryption_cache_size) | 16MB | 2 - 512  |
| [`allocator.disk_pages_encryption`](#allocatordisk_pages_encryption) | off | on/off |
| [`allocator.encryption_cache_page_size`](#allocatorencryption_cache_page_size) | 32KB |  |
| [`allocator.encryption_cache_size`](#allocatorencryption_cache_size) | 16MB |  |

## Variable descriptions

### GCache encryption

The following sections describe the variables related to GCache encryption.
All variables are read-only.

#### gcache.encryption

Enable or disable GCache cache encryption.

#### gcache.encryption_cache_page_size

The size of the GCache encryption page. The value must be multiple of the CPU page size (typically 4kB). If the value is not, the server reports an error and stops.

#### gcache.encryption_cache_size

> Every encrypted file has an `encryption.cache`, which consists of pages. Use `gcache.encryption_cache_size` to configure the `encryption.cache` size. 

Reviewer: I think we should not say `encryption.cache` as there is no such configuration parameter. This is just encryption cache (or Encryption Cache). It is configured via `gcache.encryption_cache_page_size` and `gcache.encryption_cache_size` parameters

Author: Updated.

Configure the page size in the cache with [`gcache.encryption_cache_page_size`](#gcacheencryption_cache_page_size). 

> The maximum value for the `encryption.cache` is 512 pages. This value is a hint. If the value is larger than the maximum, the value is rounded down to 512 x gcache.encryption_cache_page_size.

Reviewer: The maximum size of the Encryption Cache is...
The minimum size of the Encryption Cache is...

Author: updated.

The minimum value for the  `encryption.cache` is 2 pages. If the value is smaller, the value is rounded up.

> ### Write-Set cache encryption

Reviewer: Similar comments as for GCache encryption

Reviewer: I can't find this:
```
I think we should also explain the following things:

Rotation of GCache Master Key ('ALTER INSTANCE ROTATE GCACHE MASTER KEY')
Keyring configuration (there has to be a keyring plugin loaded, if a keyring file plugin is used, keyring file has to be located outside data directory, etc.. See the design document)
```

Author: Added.

# GCache and Write-Set cache encryption

> The feature is a tech preview. Before using this feature in production, we recommend that you test restoring production from physical backups in your environment, and also use the alternative backup method for redundancy.

Reviewer 2: Add a link to the tech preview description.

Author: Done

## GCache encryption Master Key

GCache encrypts with a File Key. Creating the galera.gcache file or a full reset generates the File Key. The File Key is encrypted with the Master Key. 

> GCache has a a RingBuffer on-disk file to manage write-sets. The file is pre-created when the server starts. This file contains a 1KB preamble to hold key-value pairs. The Master Key id and the encrypted File Key are stored in the GCache Ring preamble. The preamble is not encrypted because this section does not store sensitive data. 

Reviewer 2: ```suggestion
GCache has a RingBuffer on-disk file to manage write-sets. The file is pre-created when the server starts. This file contains a 1KB preamble to hold key-value pairs. The Master Key id and the encrypted File Key are stored in the GCache Ring preamble. The preamble is not encrypted because this section does not store sensitive data. 
```

Author: removed the extra "a"

| `enc_fk_id` | The File Key ID used for GCache ring buffer encryption. Encrypted with the Master Key and stored in base64 encoding. |
| `enc_crc` | The cyclic redundancy check (CRC) calculated from all encryption-related keys. |

### Controlling encryption

Encryption is controlled using the wsrep_provider_options. 

| Variable name | Default value | Allowed values |
|---|---|---|
| [`gcache.encryption`](#gcacheencryption) | off | on/off |
| [`gcache.encryption_cache_page_size`](#gcacheencryption_cache_page_size) | 32KB | 2-512 |
| [`gcache.encryption_cache_size`](#gcacheencryption_cache_size) | 16MB | 2 - 512  |
| [`allocator.disk_pages_encryption`](#allocatordisk_pages_encryption) | off | on/off |
| [`allocator.encryption_cache_page_size`](#allocatorencryption_cache_page_size) | 32KB |  |
| [`allocator.encryption_cache_size`](#allocatorencryption_cache_size) | 16MB |  |

## Variable descriptions

The following sections describe the variables related to Record-Set cache encryption.
All variables are read-only.

### gcache.encryption

Enable or disable GCache encryption.

### gcache.encryption_cache_page_size

> The size of the encryption cache for Record-Set pages. The value must be multiple of the CPU page size (typically 4kB).  If the value is not, the server reports an error and stops.

Reviewer 2: .  If  -> remove an extra space before If

Author: Done

The maximum limit is 512. The minimum limit is 2.

### gcache.encryption_cache_size

> The size of the encryption cache for GCache on-disk files. If `gcache.encryption=yes`, one encryption cache is associated with the GCache Ringbuffer file. Create an on-disk page and an encryption cache for that page is also created.

Reviewer 2: Create an on-disk page and an encryption cache for that page is also created. -> add comma before "and"

Author: This is not a list

The value cannot be smaller than the CPU page size, typically 4KB. If the value is smaller, the server reports an error and stops. We recommend a large cache since an extremely small encryption cache causes frequent encrypted pages swapping and degrade performance. 

The `encryption_cache_size` should be a multiple of `encryption_cache_page_size`. If the value is not, the actual cache size is rounded up.

The `encryption_cache_size` maximum is 512 allocation pages in the encryption cache. If the cache size must increase, you must also increase the `allocator.encryption_cache_page_size`.

### allocator.disk_pages_encryption

> Enable or disable the allocator's disk pages encryption.

Reviewer 2: disk page encryption

Author: should be pages

### allocator.encryption_cache_page_size

The size of the encryption cache for Record-Set pages. The value must be multiple of the CPU page size (typically 4kB).  If the value is not, the server reports an error and stops.

### allocator.encryption_cache_size

The size of the encryption cache for GCache on-disk files. If the `allocator.encryption=yes`, create an on-disk page and an encryption cache for that page is also created based on the `encryption_cache_size` value. 

> The value cannot be smaller than the CPU page size, typically 4KB. If the value is smaller, the server reports an error and stops. We recommend a large cache since an extremely small encryption cache causes frequent encrypted pages swapping and degrade performance. 

Reviewer 2: ```suggestion
The value cannot be smaller than the CPU page size, typically 4KB. If the value is smaller, the server reports an error and stops. We recommend a large cache since an extremely small encryption cache causes frequent encrypted page swapping and degraded performance. 
```

Author: pages (plural) is correct
added s for degrades to match causes

The `encryption_cache_size` should be a multiple of `encryption_cache_page_size`. If the value is not, the actual cache size is rounded up.

> The `encryption_cache_size` maximum is 512 allocation pages in the encryption cache. If the cache size must increase, you must also increase the `allocator.encryption_cache_page_size`.

Reviewer 2: If the cache size must increase -> If the cache size increases

Author: removed the "you must"

# GCache and Write-Set cache encryption

The feature is a [tech preview](../glossary.md#tech-preview). Before using this feature in production, we recommend that you test restoring production from physical backups in your environment, and also use the alternative backup method for redundancy.

## GCache encryption Master Key

GCache encrypts with a File Key. Creating the galera.gcache file or a full reset generates the File Key. The File Key is encrypted with the Master Key. 

!!! admonition "See also"

>     For more information, see [Understanding GCache and Record-set Cache](../manual/gcache_record-set_cache_difference.md)

Reviewer 2: ```suggestion
    
    For more information, see [Understanding GCache and Record-set Cache](../manual/gcache_record-set_cache_difference.md)
```

Author: Done.
